File: Code/Python Code/exceptions.py
import math

def quotientUnsafe(a:float, b:float) -> float:
    return a/b


class DivisionByZeroError(Exception):
    def __init__(self,numerator:float, denominator:float):
        self.numerator = numerator
        self.denominator = denominator

def quotient(a:float, b:float) -> float: #maybe a float
    if b == 0:
        raise DivisionByZeroError(a,b)
    else:
        return a/b

def mixedFraction(a:float, b: float, c:float) -> float: #maybe a float
    return a + quotient(b,c)



def quotientString(a:float,b:float) -> str: #always a string
    try:
        wholePart = math.floor(quotient(a,b))
        decimalPart = quotient(a,b) - wholePart
        l = []
        print(l[0])
        return str(wholePart) + " and " + str(decimalPart)
    except Exception:
        return "undefined number"

def quotientList(l:list[float],m:list[float]) -> [float]: #always a list of float
    r = []
    for i in range(0,len(l)):
        try:
            r.append(quotient(l[i],m[i]))
        except DivisionByZeroError:
            r.append(math.inf)
        except IndexError:
            pass
    return r

File: Code/Kotlin Code/exceptions.kt
import kotlin.math.floor

class DivisionByZeroError(message: String) : Exception(message)


fun quotientUnsafe(a: Int, b:Int): Int {
    return a/b
}

fun quotient(a: Float, b: Float): Float {
    when (b) {
        0f -> throw DivisionByZeroError("Division By Zero: $a/$b")
        else -> return a/b
    }
}

fun mixedFraction(wholePart: Int, numerator: Int, denominator: Int): Float {
    return wholePart + quotient(numerator.toFloat(), denominator.toFloat())
}

fun quotientString(a: Float, b: Float): String {
    try {
        val quotient = quotient(a,b)
        val wholePart = floor(quotient(a,b))
        return "whole part: $wholePart, decimal part: ${quotient - wholePart}" 
    } catch (ex: DivisionByZeroError) {
        return "undefined number"
    }
}

fun quotientList(l: List<Float>, m: List<Float>): List<Float> {
    val r: MutableList<Float> = mutableListOf()
    for (i in 0..(l.size-1)) {
        try {
            r.add(quotient(l[i], m[i]))
        } catch (ex: DivisionByZeroError) {
            r.add(Float.POSITIVE_INFINITY)
        } catch (ex: IndexOutOfBoundsException) {
            //do not append anything to the list
        }
    }
    return r.toList()
}

fun main() {
    println(quotientList(listOf(1f,2f,3f,4f,5f,6f), listOf(3f,0f,2f,2f,1f)))
}
